### DISCUSSION TITLE
making next / previous buttons without pagination but with order declared in the navigation plugin

### DISCUSSION THREAD

**Discussion by phineas-pta** [11ty/eleventy#3037] [Show and tell]
Hi all 👋 just want to show a snippet of code i created

so i have folder structure like this:
```
src/
├─ subject-1/
│  ├─ book-1/
│  │  ├─ chapter-1.html
│  │  ├─ chapter-2.html
│  ├─ book-2/
│  │  ├─ chapter-1.html
│  │  ├─ chapter-2.html
│ ...
├─ subject-2/
│  ├─ book-3/
│ ...
```
i rely totally on navigation plugin to make table of contents per subject then per book, also the breadcrumbs, so front matter for any subject - book - chapter must be:
```yaml
eleventyNavigation:
  parent: AAA
  key: BBB
  order: nnn
```

then i want to implement next / previous buttons for each chapter but `getPreviousCollectionItem` / `getNextCollectionItem` doesnt work with `eleventyNavigation` 😞

if i understand correctly official docs:
- with https://www.11ty.dev/docs/pagination/nav/ i have to add `pagination` field to front matter
- with https://www.11ty.dev/docs/filters/collection-items/ i have to add `tags` field to front matter

so like i have to repeat the order & parent title in front matter with those fields (pls correct me if im wrong)

my solution is to create a filter to get url of next / previous chapter
```js
module.exports = (eleventyConfig) => {
  eleventyConfig.addNunjucksFilter("findNextPrev", findNextPrevious)
};
function findNextPrevious(key, nodes) {
  var prev_i, next_i;
  for (let i = 0; i < nodes.length; i++) {
    // nodes already sorted by order with eleventyNavigation
    if (nodes[i]["key"] == key) {
      prev_i = i - 1;
      next_i = i + 1;
      break;
    }
  }
  return {
    "prev_url": prev_i < 0 ? null : nodes[prev_i]["url"],
    "next_url": next_i < nodes.length ? nodes[next_i]["url"] : null
  }
}
```
then i can do in a chapter (nunjucks template):
```nunjucks
{% set navPages = collections.all | eleventyNavigation(eleventyNavigation.parent) %}
{% set prev_next = eleventyNavigation.key | findNextPrev(navPages) %}

<nav>

  {% if prev_next.prev_url %}
    <a href="{{ prev_next.prev_url }}">Previous</a>
  {% else %}
    <a>🚫</a>
  {% endif %}

  {% if prev_next.next_url %}
    <a href="{{ prev_next.next_url }}">Next</a>
  {% else %}
    <a>🚫</a>
  {% endif %}

</nav>
```
im open to hear any suggestion to improve this kind-of boilerplate code 😄

**edit 1** simplify code a bit & change md to html

**Comment by pdehaan** [comment]
I'd have to try recreating this locally (unless you have a public repo I can clone), but would it work if each book was it's own collection and then you try using `getPreviousCollectionItem` / `getNextCollectionItem`? (it might need some extra re-sorting using `.addCollection()` to reorder the collection's items based on the `eleventyNavigation.order` value)

**Reply by phineas-pta** [reply]
i have a repo here: https://github.com/phineas-pta/co-han-van (warning: contain some big font files)

it's a bit messy and not entirely in english 😅 so here a bit explaination
- all the html files are in `src/` and also i set a `pathPrefix`
- for now only the `han-van-nhap-mon/` subject (1st link in homepage) has a book which named `tam-tu-kinh/` (1st link in subject page) with all the chapters (in html not markdown)
- CSS file is generated from SASS file in `assets/` so must compile it once
- JS files (not eleventyConfig) are in `src/_misc/js-*.njk` which include other JS from `_includes/*.js` 😅 

**edit 1** rename folder

**Reply by phineas-pta** [reply]
@pdehaan u mean add `.addCollection()` in 11ty config for each book folder ? with an extra sort function like `(a, b) => b.eleventyNavigation.order - a.eleventyNavigation.order;`

**Reply by pdehaan** [reply]
@phineas-pta Yeah, I was thinking something like that. But if you know which collections need to be re-sorted, you could do it in a loop so it shouldn't be too bad.
Sorry, I looked through your repo but haven't had a chance to clone it yet and take a proper look. I started working on a simple prototype, but then got stuck trying to create advanced navigation with too much magic happening in `eleventyComputed` (I think the only thing I was setting in my front matter was a page title and `eleventyNavigation.order` and almost everything else was computed.

**Reply by pdehaan** [reply]
So, I think this might work:

```njk
<!-- _includes/lay-chap.njk -->
	<nav class="pagination is-large" role="navigation" aria-label="pagination">
		{%- set bookColl = collections[book] -%}
		{%- if bookColl -%}
			{%- set previousPost = bookColl | getPreviousCollectionItem -%}
			{%- set nextPost = bookColl | getNextCollectionItem -%}
		{%- endif -%}

		{% if previousPost -%}
			<a id="prev-button" class="pagination-previous button is-primary" href="{{ previousPost.url | url }}">Previous??</a>
		{%- else -%}
			<a class="pagination-previous is-disabled">🚫</a>
		{%- endif %}

		{% if nextPost -%}
			<a id="next-button" class="pagination-next button is-primary" href="{{ nextPost.url | url }}">Next!!</a>
		{%- else -%}
			<a class="pagination-next is-disabled">🚫</a>
		{%- endif %}
	</nav>
```

```js
// src/han-van-nhap-mon/tam-tu-kinh/tam-tu-kinh.11tydata.js

const book = "tam-tu-kinh";

module.exports = {
  book,
  tags: [book],
  eleventyNavigation: {
    parent: "Tam tự kinh",
  },
};
```

Basically the src/han-van-nhap-mon/tam-tu-kinh directory data file creates a custom collection named "tam-tu-kinh" and sets a "book" front matter variable.
Then in the _includes/lay-chap.njk layout file, we get the dynamic book collection, using `collections[book]`, and then we can use the `getPreviousCollectionItem` and `getNextCollectionItem` built-in filters against our custom collection.

**Reply by pdehaan** [reply]
I was experimenting with setting `eleventyNavigation.parent` in the directory data file so you might not need to set them in each chapter-nnn.html file. You could also set the `layout` in the directory data file.

```diff
---
-layout: lay-chap.njk
eleventyNavigation:
-  parent: "Tam tự kinh"
  key: "Thái Văn Cơ"
  order: 38
---
```

We'd still need `eleventyNavigation.key` and `eleventyNavigation.order`, although it might be possible to set `order` using `eleventyComputed` based on the chapter's filename:

```diff
// src/han-van-nhap-mon/tam-tu-kinh/tam-tu-kinh.11tydata.js
const book = "tam-tu-kinh";

module.exports = {
  book,
  tags: [book],
  layout: "lay-chap.njk",
  eleventyNavigation: {
    parent: "Tam tự kinh",
  },
+  eleventyComputed: {
+    eleventyNavigation: {
+      order(data) {
+        if (data.eleventyNavigation.order) {
+          return data.eleventyNavigation.order;
+        }
+        return parseInt(data.page.fileSlug.replace("chapter-", ""), 10);
+      }
+    }
+  },
};
```

And now my chapter-002.html file front matter looks like this:
```njk
---
eleventyNavigation:
  key: "Tích Mạnh mẫu"
---
```

**Reply by phineas-pta** [reply]
tysm that's super neat, u even auto detect order from filename, what ive done is a python script to generate cahpters files with front matter pre-filled 😅

**Reply by phineas-pta** [reply]
so there should be a `.11tydata.js` per book folder ? but the content is like very much the same, u think a global data file is possible ?

**Reply by pdehaan** [reply]
> _so there should be a .11tydata.js per book folder ? but the content is like very much the same, u think a global data file is possible?_

There are possibly a few ways to do it, that was just the first/easiest thing that came to mind.
It might depend on how many books you have, but I find that a directory data file is easier than duplicating the front matter in _n_ chapters. But if you're already using Python or something else to build the chapters, maybe front matter is the better solution for your project/use case.

We could probably make it even more obscure by having a helper function that takes 1-2 arguments (book name and `eleventyNavigation.parent`, although perhaps the book/collection/tag name can be generated from a slugified version of `eleventyNavigation.parent`) and returns the rest of the contents of the .11tydata.js file. Or create an .11tydata.js file higher up (at a global "books" level, or a  "section" level) that defines the shared information (like `layout`, and `eleventyComputed.eleventyNavigation.order`) and then other directory data files just keep adding 1-2 more pieces of data specific o the book.

I'm not sure how you'd do it from a global data file easily. Even calculating `eleventyNavigation.key` or `eleventyNavigation.parent` based on filenames and depths might be too complex to be maintainable and hardcoding them somewhere is the better solution.

```js
// utils/helpers.js (WARNING: This is all just untested code, probably has some bugs):
module.exports.dirDataFile = function (parent, book) {
  return {
    book,
    tags: [book],
    layout: "lay-chap.njk",
    eleventyNavigation: {
      parent,
    },
    eleventyComputed: {
      eleventyNavigation: {
        order(data) {
          if (data.eleventyNavigation.order) {
            return data.eleventyNavigation.order;
          }
          return parseInt(data.page.fileSlug.replace("chapter-", ""), 10);
        }
      }
    },
  },
};
```

Now your book's directory data file maybe just has to import the helper function, like this untested code:
```js
const { dirDataFile } = require("../../../utils/helpers");

module.exports = dirDataFile("Tam tự kinh", "tam-tu-kinh");
```

**Reply by pdehaan** [reply]
Interesting! I can seem to slugify the... Vietnamese(?) and it "just works":

```js
console.log(eleventyConfig.getFilter("slugify")("Tam tự kinh")); // tam-tu-kinh
console.log(eleventyConfig.getFilter("slugify")("Tích Mạnh mẫu")); // tich-manh-mau
```

So maybe you _CAN_ just specify the `eleventyNavigation.parent`'s text and create the slugified `tag` for a collection name.

**Reply by phineas-pta** [reply]
tysm that's interesting, i'll look into it, that python script is just a stupid helper coz idk how to do in js properly 😂 i even wanted to put chapter contents to json then dump to html but too lazy to learn js 😂
